src/shell-app-system.c

```diff
@@ -213,11 +213,41 @@ scan_startup_wm_class_to_id (ShellAppSystem *self)
   g_list_free_full (apps, g_object_unref);
 }
 
+static gboolean
+shell_app_system_app_info_equal (GDesktopAppInfo *one,
+                                 GDesktopAppInfo *two)
+{
+  GAppInfo *one_info, *two_info;
+
+  g_return_val_if_fail (G_IS_DESKTOP_APP_INFO (one), FALSE);
+  g_return_val_if_fail (G_IS_DESKTOP_APP_INFO (two), FALSE);
+
+  one_info = G_APP_INFO (one);
+  two_info = G_APP_INFO (two);
+
+  return
+    g_app_info_equal (one_info, two_info) &&
+    g_app_info_should_show (one_info) == g_app_info_should_show (two_info) &&
+    g_strcmp0 (g_desktop_app_info_get_filename (one),
+               g_desktop_app_info_get_filename (two)) == 0 &&
+    g_strcmp0 (g_app_info_get_executable (one_info),
+               g_app_info_get_executable (two_info)) == 0 &&
+    g_strcmp0 (g_app_info_get_commandline (one_info),
+               g_app_info_get_commandline (two_info)) == 0 &&
+    strcmp (g_app_info_get_name (one_info),
+            g_app_info_get_name (two_info)) == 0 &&
+    g_strcmp0 (g_app_info_get_description (one_info),
+               g_app_info_get_description (two_info)) == 0 &&
+    strcmp (g_app_info_get_display_name (one_info),
+            g_app_info_get_display_name (two_info)) == 0 &&
```

Comment:
In understand this is just copy-pasted code from the section below, but this apparent random use of `strcmp()` and `g_strcmp0()` looks quite inconsistent. If there's an assumption that `g_app_info_get_display_name()` and `g_app_info_get_name()` are never `NULL`, there should be an assertion in those functions.

Anyway, not something to fix here, but probably worth looking at fixing upstream. Since moving these checks to `g_app_info_equal()` was [actually suggested](https://bugzilla.gnome.org/show_bug.cgi?id=773636#c8) when this code was initially introduced.

Reply:
I agree that the inconsistency makes it harder to read; I'll just make them all use `g_strcmp0()`. No harm in a few bonus `NULL` checks.

I read through those threads and I agree with desrt that changing `g_app_info_equal()`s behaviour to match this is not really an option. If one were adding a deeper equality function to GLib, it would have to be more exhaustive than this handful of checks: anything other than a full, deep comparison on the internal `GKeyFile` (expensive), or perhaps the inode number & mtime of the source files for the two (captured at construction time) would be misleading. So… I also don't like this ad-hoc list here in the shell but it's not clear exactly what should be pushed to GLib. As you say: let's think about this when we have a bit more breathing room!